File: 프로그래머스/4/1843. 사칙연산/사칙연산.js
// (1) dp[i][i] = arr[i]
// (2) dp[i][i + 1] = op[i](arr[i], arr[i + 1]) (i < arr.length - 1)
// (3) dp[i][j] = op[i](dp[i] ,dp[i + 1][j]), op[j - 1](dp[j], dp[i][j - 1]) (j = i + len - 1)
// time complexity: O(n^2) (the number of operands)
// space complexity: O(n^2)

function solution(arr) {
    const { operands, operators } = arr.reduce((accumulator, token) => {
        if (token === '+' || token === '-') {
            accumulator.operators.push(token);
            return accumulator;
        }
        accumulator.operands.push(Number(token));
        return accumulator;
    }, {
        operands: [],
        operators: []
    });
    const dp = Array.from({ length: operands.length }, () => Array.from({ length: operands.length }, () => [ -Infinity, Infinity ]));    
    
    for (let len = 1; len <= dp.length; len++) {
       for (let i = 0; i < dp.length; i++) {
           const j = i + len - 1;
           if (j >= dp.length) continue;
           if (i === j) {
               dp[i][j] = [ operands[i], operands[i], ];
               continue;
           }
           for (let k = i; k < j; k++) {
               const candidate1 = operators[k] === '+' ? dp[i][k][0] + dp[k + 1][j][0] : dp[i][k][0] - dp[k + 1][j][1];
               const candidate2 = operators[k] === '+' ? dp[i][k][1] + dp[k + 1][j][1] : dp[i][k][1] - dp[k + 1][j][0];
               dp[i][j] = [ Math.max(candidate1, dp[i][j][0]), Math.min(candidate2, dp[i][j][1])] 
           }
//            const candidate1 = operators[i] === '+' ? dp[i][i][0] + dp[i + 1][j][0] : dp[i][i][0] - dp[i + 1][j][1];
//            const candidate2 = operators[j - 1] === '+' ? dp[i][j - 1][0] + dp[j][j][0] : dp[i][j - 1][0] - dp[j][j][1];
           
        
//            const candidate3 = operators[i] === '+' ? dp[i][i][1] + dp[i + 1][j][1] : dp[i][i][1] - dp[i + 1][j][1][0];
//            const candidate4 = operators[j - 1] === '+' ? dp[i][j - 1][1] + dp[j][j][1] : dp[i][j - 1][1] - dp[j][j][0];
//            dp[i][j] = [ Math.max(candidate1, candidate2), Math.min(candidate3, candidate4) ];
       }
    }
    
    console.log(dp)
    return dp[0][dp.length - 1][0];
}

File: 프로그래머스/4/1843. 사칙연산/사칙연산.py
from math import inf

def solution(arr):
    operands = []
    operators = []
    for i in range(0, len(arr)):
        if i % 2 == 0:
            num = int(arr[i])
            operands.append(num)
        else:
            operators.append(arr[i])
    dp = [ [ [-inf, inf] for j in range(0, len(operands)) ] for i in range(0, len(operands)) ]
    
    for j in range(0, len(operands)): # length
        for i in range(0, len(operands)): # starting index
            start = i
            end = i + j
            if end >= len(operands):
                continue
                
            if start == end:
                dp[start][end][0] = operands[start]
                dp[start][end][1] = operands[start]
                continue
            
            for middle in range(start, end):
                max_result = dp[start][middle][0] + dp[middle + 1][end][0] if operators[middle] == "+" else dp[start][middle][0] - dp[middle + 1][end][1]
                min_result = dp[start][middle][1] + dp[middle + 1][end][1] if operators[middle] == "+" else dp[start][middle][1] - dp[middle + 1][end][0]
                
                dp[start][end][0] = max(dp[start][end][0], max_result)
                dp[start][end][1] = min(dp[start][end][1], min_result)
            
                    
    return dp[0][len(operands) - 1][0]
        


# dp[i][i] = arr[i] (0 <= i < 숫자의 개수)
# dp[i][j] = max(
# dp[i][i] operator dp[i + 1][j],
# dp[i][i + 1] operator dp[i + 2][j],
# ...,
# dp[i][j] operator dp[j - 1][j]) (i <= j)
# 

# 1. dp[i][i] = arr[i] 할당
# 2. for j in range: j는 길이
#      for i in range: i는 시작점 
#        for k in range(0, j + 1)
#           dp[i][j] = max(dp[i][k] operator dp[k][j])

# 틀린 부분: 합칠 때  - 인 경우 최대 최대가 아닌, 최대 / 최소가 dp[i][j]의 최댓값을 만든다.
# 따라서 각 항의 최댓값, 최솟값을 따로 구해야한다.